qiskit_experiments/framework/json.py

```diff
@@ -0,0 +1,516 @@
+        convert the value returned by ``__json_encode__`` back to the object.
+        This method should have signature
+
+        .. code-block:: python
+
+            @classmethod
+            def __json_decode__(cls, value: Any) -> cls:
+                # recover the object from the `value` returned by __json_encode__
+
+    2.  The object has a ``settings`` property. This should have signature
+
+        .. code-block:: python
+
+            @property
+            def settings(self) -> Dict[str, Any]:
+                # Return settings value for reconstructing the instance
+
+        Deserialization of objects from the ``value`` dictionary returned by
+        ``settings`` is done by calling the class `__init__` method
+        ``cls(**settings)``.
+
+    3.  In all other cases only the object class is saved. Deserialization
+        will attempt to recover the object from default initialization of
+        its class as ``cls()``.
+
+    .. note::
+
+        Serialization of custom classes works for user-defined classes in
+        Python scripts, notebooks, or third party modules. Note however
+        that these will only be able to be de-serialized if that class
+        can be imported form the same scope at the time the
+        :class:`ExperimentDecoder` is invoked.
+    """
+
+    def default(self, obj: Any) -> Any:  # pylint: disable=arguments-differ
+        if isinstance(obj, complex):
+            return _serialize_safe_float(obj)
+        if isinstance(obj, set):
+            return {"__type__": "set", "__value__": list(obj)}
+        if isinstance(obj, np.ndarray):
+            value = _serialize_and_encode(obj, np.save, allow_pickle=False)
+            return {"__type__": "ndarray", "__value__": value}
+        if isinstance(obj, sps.spmatrix):
+            value = _serialize_and_encode(obj, sps.save_npz, compress=False)
+            return {"__type__": "spmatrix", "__value__": value}
+        if isinstance(obj, bytes):
+            return _serialize_bytes(obj)
+        if dataclasses.is_dataclass(obj):
+            return _serialize_object(obj, settings=dataclasses.asdict(obj))
+        if isinstance(obj, QuantumCircuit):
+            # TODO Remove the decompose when terra 6713 is released.
+            if isinstance(obj, BlueprintCircuit):
+                obj = obj.decompose()
+            value = _serialize_and_encode(
+                data=obj, serializer=lambda buff, data: qpy_serialization.dump(data, buff)
+            )
+            return {"__type__": "QuantumCircuit", "__value__": value}
+        if isinstance(obj, ParameterExpression):
+            value = _serialize_and_encode(
+                data=obj,
+                serializer=qpy_serialization._write_parameter_expression,
+                compress=False,
+            )
+            return {"__type__": "ParameterExpression", "__value__": value}
+        if isinstance(obj, Result):
+            return {"__type__": "Result", "__value__": obj.to_dict()}
```

Comment:
IIRC this contains datetime object and we cannot directly serialize

Reply:
Good catch, there was some datetime handling in IBMQ provider I didn't copy over since it needed an extra dependency, but I guess it was there for this reason.